x-pack/metricbeat/module/gcp/billing/_meta/fields.yml

```diff
@@ -15,3 +15,22 @@
   - name: total
     type: float
     description: Total billing amount.
+  - name: sku_id
+    type: keyword
+    description: The ID of the resource used by the service.
+  - name: sku_description
+    type: text
+    description: A description of the resource type used by the service. For example, a resource type for Cloud Storage is Standard Storage US.
+  - name: service_id
+    type: keyword
+    description: The ID of the service that the usage is associated with.
+  - name: service_description
+    type: text
+    description: The Google Cloud service that reported the Cloud Billing data.
+  - name: tags
+    type: object
+    object_type: keyword
+    description: A collection of key-value pairs that provide additional metadata.
```

Comment:
One small problem I see is if we try to filter the dashboard by tags. We probably can't use the `object` type because we don't know the fields beforehand. What if we use something like `nested` instead? That way, the field definitions would look like this:

```yaml
- name: tags
  type: nested
  fields:
  - name: key
    type: keyword
  - name: value
    type: keyword
```

I'm wondering if this is a good option so the dashboard can be filtered by tags keys and values. If yes, then the `createTagsMap` function must be changed because `nested` types expect a JSON array.

Reply:
IIIRC, Rajvi had tried out visualizations with object fields.
cc: @rajvi-elastic

Reply:
Yes @ishleenk17, Filters can be applied on both object and nested fields. It won't throw any errors.

Reply:
What I'm trying to highlight is that with the current implementation where `tags` is a map, we can't create input control groups to filter dashboard data using tags keys or tags values fields.

<b>Current implementation:</b>

```json
{
  "gcp": {
    "billing": {
      ...
      "tags": {
        "k1": "v1",
        "k2": "v2"
      }
    }
  }
}
```

<b>Storing tags as an array of nested tag objects:</b>

```json
{
  "gcp": {
    "billing": {
      ...
      "tags": [
        {
          "value": "v1",
          "key": "k1"
        },
        {
          "value": "v2",
          "key": "k2"
        }
      ]
    }
  }
}
```

The problem with using a map is that in an input control group you can't use `tags.key` or `tags.value` as a field because the mapping is "dynamic" and you can't know the tags beforehand.

![gcp_billing_control](https://github.com/elastic/beats/assets/94497545/accac0dd-4b62-46db-8a0b-a26000a206d0)

Using `nested` type allows us to filter by `gcp.billing.tags.key` or `gcp.billing.tags.value`.

![gcp_billing_control_updated](https://github.com/elastic/beats/assets/94497545/20995046-bf95-4def-a5f3-05fcb591f2ea)

I'm unsure if filtering by `key` and `value` separately is ideal, or if it should be done by the combination of `key+value`.

Reply:
As per me a concat of key+value makes more sense for filtering, rather than individual filtering on just key or just value. 
@kaiyan-sheng @zmoog any thoughts on this since you both have worked on billing modules of other cloud packages.

Reply:
Lets have the key and value as separate fields.
If need be we can have the key+value as a concat in the pipeline. and use that for filtering in dashboards.